proposals/819-kubeflow-sdk/README.md

```diff
@@ -0,0 +1,423 @@
+
+## Summary
+
+Original Google document: https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit
+
+This KEP outlines a proposal to design Kubeflow SDK that streamlines user experience for
+data scientists and ML engineers to interact with various Kubeflow projects
+(Trainer, Katib, Spark Operator, Pipelines, Model Registry, KServe).
+
+## Motivation
+
+Kubeflow users come from diverse backgrounds, including cluster administrators, data scientists,
+and ML engineers (referred to as `ML User` in this proposal). While administrators are familiar
+with Kubernetes, ML Users typically are not. They prefer working in a familiar Python environment
+with popular data and ML frameworks like PyTorch, TensorFlow, and JAX for ML and Ray, Spark,
+Pandas, and Dask for data, while leveraging advanced hardware and cloud infrastructure
+to scale their experiments.
+
+Typically, ML Users want to:
+
+- Open JupyterLab.
+- Install required AI/ML libraries with `pip`.
+- Prototype ML code using PyTorch/TensorFlow APIs.
+- Launch jobs to create features from raw data and prepare training datasets.
+- Launch jobs to train, fine-tune, and optimize models across many GPUs.
+- Collect and analyze results within the same JupyterLab environment.
+
+One of the typical ML experience workflows can be seen at
+[this KubeCon + CloudNativeCon talk](https://youtu.be/Lgy4ir1AhYw?t=454).
+
+Currently, Kubeflow's ecosystem lacks simple interfaces for interacting with its components. This
+complexity hinders adoption among ML Users, as it requires Kubernetes expertise to use effectively.
+
+This document proposes the Kubeflow Python SDK, designed to streamline the experience of
+ML users when interacting with Kubeflow components. In the future, we may extend this SDK to
+support additional languages (e.g., Swift, Rust, Java) based on user demand.
+
+### Goals
+
+- Collaborate with ML Experience/IDE Working Group to include this SDK into their charter.
+
+- Introduce initial version of Kubeflow Python SDK that supports Kubeflow Trainer and Kubeflow
+  Katib APIs.
+
+  - **Note** that Kubeflow Katib will be probably renamed to Kubeflow Optimizer in the future.
+    Thus, in this proposal we will keep the Kubeflow Optimizer name moving forward.
+
+- Reduce duplication across `kubeflow-training` and `kubeflow-katib` SDKs by reusing code assets.
+
+- Design the SDK version control and release schedule.
+
+- Design synchronization scripts between Kubeflow control plane and SDK models.
+
+### No Goals
+
+- Consolidate Kubeflow Pipelines APIs into centralized Kubeflow SDK.
+
+  - As we discussed in the Google document, we want to postpone the `kfp` integration after
+    we release the initial version of Kubeflow SDK. The Kubeflow community will work towards
+    long-term goal to provide consistent experience for ML Users to interact with Kubeflow APIs.
+
+- Integrate Kubeflow Spark Operator, KServe, Model Registry into Kubeflow SDK.
```

Comment:
if this is about not-integrating installations, controller-handling, etc SGTM, but is this meant to not considering 

```
such as providing a mechanism to load existing clients as “extras” (such as the case for existing clients as MR) and/or Plug-in/Provider mechanism (ie provide a high level interface which existing clients can implement
```

which was discussed per requested [feedback in the doc](https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit?tab=t.0#heading=h.f6k29hmtpp6m) please for my understanding?

Reply:
Yes, I think we should talk about integration of existing clients into Kubeflow SDK in a followup KEP/discussions.
We should discuss what is the right approach for us to integrate existing client (e.g. model-registry) into Kubeflow SDK.
For example, for `kfp`, we can consolidate client APIs into `PipelinesClient()` as @chensun suggested here: https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit?disco=AAABdUN58Ss